Add dotnet-trace to Debug ThreadPool Starvation (#46251)
dotnet/docs: docs/core/diagnostics/debug-threadpool-starvation.md

# Debug ThreadPool Starvation

> **This article applies to: ✔️** .NET 9.0 and later versions

Author: Makes sense, done! I wasn't able to copy the formatting from the terminal though.

Press p to pause, r to resume, q to quit.
    Status: Running

> Name Current Value

Reviewer: Can you align the 'Current Value' column so that it visually matches what you'd see in the tool?

Author: Done

```

ThreadPool starvation occurs when there are no free threads to handle the queued work items and the runtime responds by increasing the number of ThreadPool threads. You should observe the `dotnet.thread_pool.thread.count` rapidly increase to 2-3x the number of processor cores on your machine and then further threads are added 1-2 per second until eventually stabilizing somewhere above 125. The slow and steady increase of ThreadPool threads combined with CPU Usage much less than 100% are the key signals that ThreadPool starvation is currently a performance bottleneck. The thread count increase will continue until either the pool hits the maximum number of threads, enough threads have been created to satisfy all the incoming work items, or the CPU has been saturated. Often, but not always, ThreadPool starvation will also show large values for `dotnet.thread_pool.queue.length` and low values for `dotnet.thread_pool.work_item.count`, meaning that there's a large amount of pending work and little work being completed. Here's an example of the counters while the thread count is still rising:

```dotnetcli
[System.Runtime]

>     dotnet.assembly.count ({assembly}) 115

Author: Done! Turns out llm can do this.

## Resolving ThreadPool starvation

> To eliminate ThreadPool starvation, ThreadPool threads need to remain unblocked so that they're available to handle incoming work items. There are multiple ways to determine what each thread was doing: using the [dotnet-stack](dotnet-stack.md) tool, capturing a dump with [dotnet-dump](dotnet-dump.md) that can be viewed in [Visual Studio](/visualstudio/debugger/using-dump-files), or collecting wait events with [dotnet-trace](dotnet-trace.md). dotnet-stack can be faster because it shows the thread stacks immediately on the console, but Visual Studio dump debugging offers better visualizations that map frames to source, Just My Code can filter out runtime implementation frames, and the Parallel Stacks feature can help group large numbers of threads with similar stacks. This tutorial first shows the dotnet-stack option, then demonstrates how to use dotnet-trace. See the [diagnosing ThreadPool starvation tutorial video](/shows/on-net/diagnosing-thread-pool-exhaustion-issues-in-net-core-apps) for an example of investigating the thread stacks using Visual Studio.

Reviewer: ```suggestion
To eliminate ThreadPool starvation, ThreadPool threads need to remain unblocked so that they're available to handle incoming work items. There are multiple ways to determine what each thread was doing. If the issue occurs only occasionally then collecting a trace with [dotnet-trace](dotnet-trace.md) is best to record application behavior over a period of time. If the issue is constantly occurring then you can use the [dotnet-stack](dotnet-stack.md) tool or capture a dump with [dotnet-dump](dotnet-dump.md) that can be viewed in [Visual Studio](/visualstudio/debugger/using-dump-files). dotnet-stack can be faster because it shows the thread stacks immediately on the console, but Visual Studio dump debugging offers better visualizations that map frames to source, Just My Code can filter out runtime implementation frames, and the Parallel Stacks feature can help group large numbers of threads with similar stacks. This tutorial shows the dotnet-stack and dotnet-trace options. See the [diagnosing ThreadPool starvation tutorial video](/shows/on-net/diagnosing-thread-pool-exhaustion-issues-in-net-core-apps) for an example of investigating the thread stacks using Visual Studio.
```

> ### dotnet-stack

Reviewer: ```suggestion
### Diagnose a continuous issue with dotnet-stack
```

  DiagnosticScenarios!testwebapi.Controllers.DiagScenarioController.TaskWait()
```

> ### dotnet-trace

Reviewer: ```suggestion
### Diagnose an intermittent issue with dotnet-trace
```

The dotnet-stack approach is effective only for obvious, consistent blocking operations that occur in every request. In some scenarios, the blocking could happen sporadically only every few minutes, making dotnet-stack less useful for diagnosing the issue. dotnet-trace can be used to collect events over a period of time and save them in a nettrace file that can be analyzed later.

> There is one particular event that helps diagnosing thread pool starvation that was introduced in .NET 9, it's the WaitHandleWait event. It's emitted when a thread pool thread becomes blocked by operations such as sync-over-async calls (e.g. `Task.Result`, `Task.Wait`, `Task.GetAwaiter().GetResult()`) or by other locking operations like `lock`, `Monitor.Enter`, `ManualResetEventSlim.Wait`, or `SemaphoreSlim.Wait`.

Reviewer: ```suggestion
There is one particular event that helps diagnosing thread pool starvation that was introduced in .NET 9, the WaitHandleWait event. It's emitted when a thread becomes blocked by operations such as sync-over-async calls (e.g. `Task.Result`, `Task.Wait`, `Task.GetAwaiter().GetResult()`) or by other locking operations like `lock`, `Monitor.Enter`, `ManualResetEventSlim.Wait`, or `SemaphoreSlim.Wait`.
```

Run Bombardier again to put the web server under load:

```dotnetcli
bombardier-windows-amd64.exe https://localhost:5001/api/diagscenario/taskwait -d 120s
```

Then run dotnet-trace to collect wait events:

```dotnetcli
dotnet trace collect -n DiagnosticScenarios --clrevents waithandle --clreventlevel verbose --duration 00:00:30
```

That should generate a file named `DiagnosticScenarios.exe_yyyyddMM_hhmmss.nettrace` containing the events. This nettrace can be analyzed using two different tools:
- [PerfView](https://github.com/microsoft/perfview/releases): a performance analysis tool developed by Microsoft for Windows only
- [.NET Events Viewer](https://verdie-g.github.io/dotnet-events-viewer): a nettrace analysis [Blazor](https://dotnet.microsoft.com/en-us/apps/aspnet/web-apps/blazor) web tool developed by the community

Let's see how to use each tool to read the nettrace file.

> ### Analyze a nettrace with Perfview

Reviewer: ```suggestion
#### Analyze a nettrace with Perfview
```

![Screenshot of the any stacks view in PerfView](./media/perfview-any-stacks.png)
4. Double click on the line "Event Microsoft-Windows-DotNETRuntime/WaitHandleWait/Start"
5. Now you should see the stack traces where the WaitHandleWait events were emitted. They are split by "WaitSource". Currently there are two sources. MonitorWait for events emitted through [Monitor.Wait](https://learn.microsoft.com/en-us/dotnet/api/system.threading.monitor), and Unknown for all the others.
![Screenshot of the any stacks view for wait events in PerfView](./media/perfview-any-stacks-waithandle.png)
6. Let's start with the MonitorWait as it represents 64.8% of the events. You can tick the checkboxes to expand the stack traces responsible for emitting this event.
![Screenshot of the expanded any stacks view for wait events in PerfView](./media/perfview-any-stacks-waithandle-expanded.png)
7. This stack trace can be read as: `Task<T>.Result` emitted a WaitHandleWait event with a WaitSource MonitorWait (`Task<T>.Result` uses `Monitor.Wait` to perform a wait). It was called by `DiagScenarioController.TaskWait`, which was called by some lambda, which was called by some ASP.NET code

> ### Analyze a nettrace with .NET Events Viewer

Reviewer: ```suggestion
#### Analyze a nettrace with .NET Events Viewer
```

> 1. Go to [verdie-g.github.io/dotnet-events-viewer](https://verdie-g.github.io/dotnet-events-viewer)

Reviewer: Neat tool!

![Screenshot of the any stacks view for wait events in PerfView](./media/perfview-any-stacks-waithandle.png)
6. Let's start with the MonitorWait as it represents 64.8% of the events. You can tick the checkboxes to expand the stack traces responsible for emitting this event.
![Screenshot of the expanded any stacks view for wait events in PerfView](./media/perfview-any-stacks-waithandle-expanded.png)
7. This stack trace can be read as: `Task<T>.Result` emitted a WaitHandleWait event with a WaitSource MonitorWait (`Task<T>.Result` uses `Monitor.Wait` to perform a wait). It was called by `DiagScenarioController.TaskWait`, which was called by some lambda, which was called by some ASP.NET code

#### Analyze a nettrace with .NET Events Viewer

1. Go to [verdie-g.github.io/dotnet-events-viewer](https://verdie-g.github.io/dotnet-events-viewer)
2. Drag-and-drop the nettrace file

![Screenshot of the opening of a nettrace in .NET Events Viewer](./media/dotnet-events-viewer-open-nettrace.png)

3. Go to the page "Events Tree", select the event "WaitHandleWaitStart", click "Run query"

> ![Screenshot of an events query in .NET Events Viewer](./media/dotnet-events-viewer-query-wait-events.png)

Author: > and then select **Run query**

Does "select" for a button make sense?

```dotnetcli
bombardier-windows-amd64.exe https://localhost:5001/api/diagscenario/taskwait -d 120s
```

Then run dotnet-trace to collect wait events:

```dotnetcli
dotnet trace collect -n DiagnosticScenarios --clrevents waithandle --clreventlevel verbose --duration 00:00:30
```

That should generate a file named `DiagnosticScenarios.exe_yyyyddMM_hhmmss.nettrace` containing the events. This nettrace can be analyzed using two different tools:

- [PerfView](https://github.com/microsoft/perfview/releases): A performance analysis tool developed by Microsoft for Windows only.
- [.NET Events Viewer](https://verdie-g.github.io/dotnet-events-viewer): A nettrace analysis [Blazor](https://dotnet.microsoft.com/apps/aspnet/web-apps/blazor) web tool developed by the community.

The following sections show how to use each tool to read the nettrace file.

#### Analyze a nettrace with Perfview

1. Download [PerfView](https://github.com/microsoft/perfview/releases) and run it.
1. Open the nettrace file by double-clicking it.
   ![Screenshot of the opening of a nettrace in PerfView](./media/perfview-open-nettrace.png)
1. Double click on **Advanced Group** > **Any Stacks**. A new window opens.

>    ![Screenshot of the any stacks view in PerfView](./media/perfview-any-stacks.png)

Author: I've committed that suggestion. Would you like me to do the same for other images? I'm not sure how to test that custom syntax on my side.

>    ![Screenshot of the any stacks view in PerfView](./media/perfview-any-stacks.png)

Reviewer 2: No, I've already gone ahead and added a lightbox to the other images that I think can benefit from it.
